File: services/management-api/internal/domain/gateway/entity/gateway.go
// Package entity contains domain entities for the Gateway bounded context
package entity

import (
	"errors"
	"net"
	"time"
)

// Common domain errors
var (
	ErrInvalidIPAddress     = errors.New("invalid IP address format")
	ErrInvalidGatewayType   = errors.New("invalid gateway type")
	ErrInvalidCarrierName   = errors.New("carrier name cannot be empty")
	ErrGatewayAlreadyActive = errors.New("gateway is already active")
	ErrGatewayNotActive     = errors.New("gateway is not active")
	ErrGatewayBlacklisted   = errors.New("gateway is blacklisted")
)

// GatewayType represents the type of network gateway
type GatewayType string

const (
	GatewayTypeLocal         GatewayType = "local"
	GatewayTypeInternational GatewayType = "international"
	GatewayTypeTransit       GatewayType = "transit"
)

// IsValid checks if the gateway type is valid
func (t GatewayType) IsValid() bool {
	switch t {
	case GatewayTypeLocal, GatewayTypeInternational, GatewayTypeTransit:
		return true
	}
	return false
}

// Gateway is the aggregate root for gateway management
type Gateway struct {
	id                string
	name              string
	ipAddress         net.IP
	carrierName       string
	gatewayType       GatewayType
	fraudThreshold    float64
	cpmLimit          int
	acdThreshold      float64
	isActive          bool
	isBlacklisted     bool
	blacklistReason   string
	blacklistExpiresAt *time.Time
	createdAt         time.Time
	updatedAt         time.Time
	
	// Read-only statistics (populated when needed)
	callsToday int
	fraudCount int
}

// NewGateway creates a new Gateway with validation
func NewGateway(id, name, ipAddress, carrierName string, gatewayType GatewayType) (*Gateway, error) {
	ip := net.ParseIP(ipAddress)
	if ip == nil {
		return nil, ErrInvalidIPAddress
	}
	
	if !gatewayType.IsValid() {
		return nil, ErrInvalidGatewayType
	}
	
	if carrierName == "" {
		return nil, ErrInvalidCarrierName
	}
	
	now := time.Now()
	return &Gateway{
		id:             id,
		name:           name,
		ipAddress:      ip,
		carrierName:    carrierName,
		gatewayType:    gatewayType,
		fraudThreshold: 0.8,  // Default 80%
		cpmLimit:       60,   // Default 60 CPM
		acdThreshold:   10.0, // Default 10 seconds
		isActive:       true,
		isBlacklisted:  false,
		createdAt:      now,
		updatedAt:      now,
	}, nil
}

// Reconstitute creates a Gateway from persisted data (no validation)
func Reconstitute(
	id, name string,
	ipAddress net.IP,
	carrierName string,
	gatewayType GatewayType,
	fraudThreshold float64,
	cpmLimit int,
	acdThreshold float64,
	isActive, isBlacklisted bool,
	blacklistReason string,
	blacklistExpiresAt *time.Time,
	createdAt, updatedAt time.Time,
) *Gateway {
	return &Gateway{
		id:                 id,
		name:               name,
		ipAddress:          ipAddress,
		carrierName:        carrierName,
		gatewayType:        gatewayType,
		fraudThreshold:     fraudThreshold,
		cpmLimit:           cpmLimit,
		acdThreshold:       acdThreshold,
		isActive:           isActive,
		isBlacklisted:      isBlacklisted,
		blacklistReason:    blacklistReason,
		blacklistExpiresAt: blacklistExpiresAt,
		createdAt:          createdAt,
		updatedAt:          updatedAt,
	}
}

// === Getters ===

func (g *Gateway) ID() string               { return g.id }
func (g *Gateway) Name() string             { return g.name }
func (g *Gateway) IPAddress() net.IP        { return g.ipAddress }
func (g *Gateway) CarrierName() string      { return g.carrierName }
func (g *Gateway) GatewayType() GatewayType { return g.gatewayType }
func (g *Gateway) FraudThreshold() float64  { return g.fraudThreshold }
func (g *Gateway) CPMLimit() int            { return g.cpmLimit }
func (g *Gateway) ACDThreshold() float64    { return g.acdThreshold }
func (g *Gateway) IsActive() bool           { return g.isActive }
func (g *Gateway) IsBlacklisted() bool      { return g.isBlacklisted }
func (g *Gateway) BlacklistReason() string  { return g.blacklistReason }
func (g *Gateway) CreatedAt() time.Time     { return g.createdAt }
func (g *Gateway) UpdatedAt() time.Time     { return g.updatedAt }
func (g *Gateway) CallsToday() int          { return g.callsToday }
func (g *Gateway) FraudCount() int          { return g.fraudCount }

// IsInternational returns true if this is an international gateway
func (g *Gateway) IsInternational() bool {
	return g.gatewayType == GatewayTypeInternational
}

// === Behavior ===

// UpdateThresholds modifies detection thresholds with validation
func (g *Gateway) UpdateThresholds(fraudThreshold *float64, cpmLimit *int, acdThreshold *float64) {
	if fraudThreshold != nil {
		t := *fraudThreshold
		if t < 0 {
			t = 0
		} else if t > 1 {
			t = 1
		}
		g.fraudThreshold = t
	}
	if cpmLimit != nil {
		c := *cpmLimit
		if c < 1 {
			c = 1
		} else if c > 1000 {
			c = 1000
		}
		g.cpmLimit = c
	}
	if acdThreshold != nil {
		a := *acdThreshold
		if a < 1 {
			a = 1
		}
		g.acdThreshold = a
	}
	g.updatedAt = time.Now()
}

// UpdateName changes the gateway name
func (g *Gateway) UpdateName(name string) {
	if name != "" {
		g.name = name
		g.updatedAt = time.Now()
	}
}

// UpdateCarrier changes the carrier name
func (g *Gateway) UpdateCarrier(carrierName string) {
	if carrierName != "" {
		g.carrierName = carrierName
		g.updatedAt = time.Now()
	}
}

// Activate enables the gateway
func (g *Gateway) Activate() error {
	if g.isActive {
		return ErrGatewayAlreadyActive
	}
	if g.isBlacklisted {
		return ErrGatewayBlacklisted
	}
	g.isActive = true
	g.updatedAt = time.Now()
	return nil
}

// Deactivate disables the gateway
func (g *Gateway) Deactivate() error {
	if !g.isActive {
		return ErrGatewayNotActive
	}
	g.isActive = false
	g.updatedAt = time.Now()
	return nil
}

// Blacklist adds the gateway to blacklist
func (g *Gateway) Blacklist(reason string, expiresAt *time.Time) {
	g.isBlacklisted = true
	g.blacklistReason = reason
	g.blacklistExpiresAt = expiresAt
	g.isActive = false
	g.updatedAt = time.Now()
}

// Unblacklist removes the gateway from blacklist
func (g *Gateway) Unblacklist() {
	g.isBlacklisted = false
	g.blacklistReason = ""
	g.blacklistExpiresAt = nil
	g.updatedAt = time.Now()
}

// SetStatistics sets read-only call statistics
func (g *Gateway) SetStatistics(callsToday, fraudCount int) {
	g.callsToday = callsToday
	g.fraudCount = fraudCount
}

// ExceedsCPMLimit checks if a given CPM exceeds the limit
func (g *Gateway) ExceedsCPMLimit(cpm int) bool {
	return cpm > g.cpmLimit
}

// IsACDSuspicious checks if average call duration is suspiciously low
func (g *Gateway) IsACDSuspicious(acd float64) bool {
	return acd < g.acdThreshold
}

File: voxguard/detection-service-rust/src/domain/aggregates/gateway.rs
//! Gateway Aggregate - Represents a network gateway/carrier
//!
//! Gateways are sources of SIP traffic. They have configurable thresholds
//! and can be blacklisted for fraud.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::domain::{
    errors::DomainError,
    events::GatewayBlockedEvent,
    value_objects::IPAddress,
};

/// Gateway type classification
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum GatewayType {
    /// Local Nigerian carrier
    Local,
    /// International gateway
    International,
    /// Transit/interconnect
    Transit,
}

/// Gateway aggregate root
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Gateway {
    /// Unique identifier
    id: Uuid,
    /// Human-readable name
    name: String,
    /// Gateway IP address
    ip_address: IPAddress,
    /// Carrier name
    carrier_name: String,
    /// Gateway classification
    gateway_type: GatewayType,
    /// Fraud detection threshold
    fraud_threshold: f64,
    /// Calls per minute limit
    cpm_limit: u32,
    /// Average call duration threshold (seconds)
    acd_threshold: f64,
    /// Whether gateway is active
    is_active: bool,
    /// Whether gateway is blacklisted
    is_blacklisted: bool,
    /// Blacklist expiration (if temporary)
    blacklist_expires_at: Option<DateTime<Utc>>,
    /// Blacklist reason
    blacklist_reason: Option<String>,
    /// Creation timestamp
    created_at: DateTime<Utc>,
    /// Last update timestamp
    updated_at: DateTime<Utc>,
}

impl Gateway {
    /// Creates a new Gateway
    pub fn new(
        name: impl Into<String>,
        ip_address: IPAddress,
        carrier_name: impl Into<String>,
        gateway_type: GatewayType,
    ) -> Self {
        let now = Utc::now();
        Self {
            id: Uuid::new_v4(),
            name: name.into(),
            ip_address,
            carrier_name: carrier_name.into(),
            gateway_type,
            fraud_threshold: 0.8,  // Default 80%
            cpm_limit: 60,         // Default 60 CPM
            acd_threshold: 10.0,   // Default 10 seconds
            is_active: true,
            is_blacklisted: false,
            blacklist_expires_at: None,
            blacklist_reason: None,
            created_at: now,
            updated_at: now,
        }
    }

    // === Getters ===

    pub fn id(&self) -> Uuid {
        self.id
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn ip_address(&self) -> &IPAddress {
        &self.ip_address
    }

    pub fn carrier_name(&self) -> &str {
        &self.carrier_name
    }

    pub fn gateway_type(&self) -> GatewayType {
        self.gateway_type
    }

    pub fn fraud_threshold(&self) -> f64 {
        self.fraud_threshold
    }

    pub fn cpm_limit(&self) -> u32 {
        self.cpm_limit
    }

    pub fn acd_threshold(&self) -> f64 {
        self.acd_threshold
    }

    pub fn is_active(&self) -> bool {
        self.is_active
    }

    pub fn is_blacklisted(&self) -> bool {
        // Check if blacklist has expired
        if self.is_blacklisted {
            if let Some(expires) = self.blacklist_expires_at {
                if expires < Utc::now() {
                    return false;
                }
            }
        }
        self.is_blacklisted
    }

    pub fn is_international(&self) -> bool {
        self.gateway_type == GatewayType::International
    }

    // === Behavior ===

    /// Updates detection thresholds
    pub fn update_thresholds(
        &mut self,
        fraud_threshold: Option<f64>,
        cpm_limit: Option<u32>,
        acd_threshold: Option<f64>,
    ) {
        if let Some(ft) = fraud_threshold {
            self.fraud_threshold = ft.clamp(0.0, 1.0);
        }
        if let Some(cpm) = cpm_limit {
            self.cpm_limit = cpm.min(1000);
        }
        if let Some(acd) = acd_threshold {
            self.acd_threshold = acd.max(1.0);
        }
        self.updated_at = Utc::now();
    }

    /// Activates the gateway
    pub fn activate(&mut self) {
        self.is_active = true;
        self.updated_at = Utc::now();
    }

    /// Deactivates the gateway
    pub fn deactivate(&mut self) {
        self.is_active = false;
        self.updated_at = Utc::now();
    }

    /// Blacklists the gateway
    pub fn blacklist(
        &mut self,
        reason: impl Into<String>,
        expires_at: Option<DateTime<Utc>>,
    ) -> GatewayBlockedEvent {
        self.is_blacklisted = true;
        self.blacklist_reason = Some(reason.into());
        self.blacklist_expires_at = expires_at;
        self.is_active = false;
        self.updated_at = Utc::now();

        GatewayBlockedEvent::new(
            &self.ip_address,
            self.blacklist_reason.as_deref().unwrap_or("Unknown"),
            expires_at,
        )
    }

    /// Removes gateway from blacklist
    pub fn unblacklist(&mut self) -> Result<(), DomainError> {
        if !self.is_blacklisted {
            return Err(DomainError::InvariantViolation(
                "Gateway is not blacklisted".into(),
            ));
        }

        self.is_blacklisted = false;
        self.blacklist_reason = None;
        self.blacklist_expires_at = None;
        self.updated_at = Utc::now();
        Ok(())
    }

    /// Checks if a CPM value exceeds the limit
    pub fn exceeds_cpm_limit(&self, cpm: u32) -> bool {
        cpm > self.cpm_limit
    }

    /// Checks if ACD is suspiciously low
    pub fn is_acd_suspicious(&self, acd: f64) -> bool {
        acd < self.acd_threshold
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::events::DomainEvent;

    fn create_test_gateway() -> Gateway {
        let ip = IPAddress::new("10.0.0.1").unwrap();
        Gateway::new("Lagos Gateway 1", ip, "MTN Nigeria", GatewayType::Local)
    }

    #[test]
    fn test_gateway_creation() {
        let gw = create_test_gateway();

        assert!(gw.is_active());
        assert!(!gw.is_blacklisted());
        assert_eq!(gw.gateway_type(), GatewayType::Local);
        assert!(!gw.is_international());
    }

    #[test]
    fn test_threshold_updates() {
        let mut gw = create_test_gateway();

        gw.update_thresholds(Some(0.9), Some(100), Some(5.0));

        assert_eq!(gw.fraud_threshold(), 0.9);
        assert_eq!(gw.cpm_limit(), 100);
        assert_eq!(gw.acd_threshold(), 5.0);
    }

    #[test]
    fn test_threshold_clamping() {
        let mut gw = create_test_gateway();

        // Values are clamped to valid ranges
        gw.update_thresholds(Some(1.5), Some(5000), Some(0.1));

        assert_eq!(gw.fraud_threshold(), 1.0);
        assert_eq!(gw.cpm_limit(), 1000);
        assert_eq!(gw.acd_threshold(), 1.0);
    }

    #[test]
    fn test_blacklist() {
        let mut gw = create_test_gateway();

        let event = gw.blacklist("Fraud detected", None);

        assert!(gw.is_blacklisted());
        assert!(!gw.is_active());
        assert_eq!(event.event_type(), "GatewayBlocked");
    }

    #[test]
    fn test_temporary_blacklist_expiry() {
        let mut gw = create_test_gateway();

        // Blacklist with past expiration
        let expired = Utc::now() - chrono::Duration::hours(1);
        gw.blacklist("Test", Some(expired));

        // Should report as not blacklisted since it expired
        assert!(!gw.is_blacklisted());
    }

    #[test]
    fn test_cpm_limit_check() {
        let gw = create_test_gateway();

        assert!(!gw.exceeds_cpm_limit(50));
        assert!(gw.exceeds_cpm_limit(100));
    }

    #[test]
    fn test_acd_suspicious_check() {
        let gw = create_test_gateway();

        // Default threshold is 10 seconds
        assert!(gw.is_acd_suspicious(5.0));  // Short calls are suspicious
        assert!(!gw.is_acd_suspicious(30.0)); // Normal call duration
    }
}
